modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/L0Service.scala

```diff
@@ -36,56 +40,87 @@ object L0Service {
 
       private val logger = Slf4jLogger.getLogger[F]
 
-      def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalSnapshot]]] =
-        l0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
-          l0GlobalSnapshotClient
-            .get(ordinal)(l0Peer)
-            .flatMap(_.toHashedWithSignatureCheck)
-            .flatMap(_.liftTo[F])
-            .map(_.some)
-        }.handleErrorWith { e =>
-          logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalSnapshot]])
+      def fetchOrdinals(peer: L0Peer): F[List[SnapshotOrdinal]] =
+        lastSnapshotStorage.getOrdinal.flatMap { lastStoredOrdinal =>
+          l0GlobalSnapshotClient.getLatestOrdinal
+            .run(peer)
+            .map { latestOrdinal =>
+              val nextOrdinal = lastStoredOrdinal.map(_.next).getOrElse(latestOrdinal).value.value
+              val latestOrdinalCap = latestOrdinal.value.value
+                .min(singlePullLimit.map(nextOrdinal + _.value).getOrElse(latestOrdinal.value.value))
+
+              nextOrdinal to latestOrdinalCap
+            }
         }
+          .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
 
-      def pullGlobalSnapshots: F[List[Hashed[GlobalSnapshot]]] = {
-        for {
-          lastStoredOrdinal <- lastSnapshotStorage.getOrdinal
-          l0Peer <- l0ClusterStorage.getRandomPeer
+      def fetchMajorityGlobalSnapshot(ordinal: SnapshotOrdinal, peers: List[L0Peer]): F[Option[Hashed[GlobalSnapshot]]] =
+        peers
+          .traverse(fetchGlobalSnapshot(ordinal))
+          .map(_.flatten)
+          .map(_.sortBy(-_.signed.proofs.length).headOption)
 
-          pulled <- l0GlobalSnapshotClient.getLatestOrdinal
-            .run(l0Peer)
-            .map { lastOrdinal =>
-              val nextOrdinal = lastStoredOrdinal.map(_.next).getOrElse(lastOrdinal)
-              val lastOrdinalCap = lastOrdinal.value.value
-                .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+      def fetchGlobalSnapshot(ordinal: SnapshotOrdinal)(peer: L0Peer): F[Option[Hashed[GlobalSnapshot]]] =
+        l0GlobalSnapshotClient
+          .get(ordinal)(peer)
+          .flatMap(_.toHashedWithSignatureCheck)
+          .flatMap(_.liftTo[F])
+          .map(_.some)
+          .handleError(_ => none[Hashed[GlobalSnapshot]])
```

Comment:
Don't we want to have it logged in case it fails?

Reply:
actually I can add logger on some global level, but imo it's not worthing to add for every single fetch 🤔 wdyt @TheMMaciek ?

Reply:
Well the way I was using that function is that first the node was asked what ordinals it has and after that these ordinals were fetched, so a failure here was unexpected hence logging. But If you have some idea on how to log it differently without loosing context then I would go for it.

Reply:
I've added logs :+1: